### PR TITLE
feat: add sync button to execute syncer once

### PR DESCRIPTION
**Pull request by seriouszyx**
image

Signed-off-by: Yixiang Zhao <seriouszyx@foxmail.com>

### CODE REVIEW

File: controllers/syncer.go [hunk 1]
```diff
@@ -114,3 +114,21 @@ func (c *ApiController) DeleteSyncer() {
 	c.Data["json"] = wrapActionResponse(object.DeleteSyncer(&syncer))
 	c.ServeJSON()
 }
+
+// @Title SyncUsers
+// @Tag Syncer API
+// @Description sync users
+// @Param   body    body   object.Syncer  true        "The details of the syncer"
+// @Success 200 {object} controllers.Response The Response object
+// @router /sync-users [post]
```
Comment (hsluoyz) [inline controllers/syncer.go:123]: /run-syncer

File: controllers/syncer.go [hunk 2]
```diff
@@ -114,3 +114,21 @@ func (c *ApiController) DeleteSyncer() {
...
+func (c *ApiController) SyncUsers() {
+	var syncer object.Syncer
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &syncer)
```
Comment (hsluoyz) [inline controllers/syncer.go:126]: Use id as param, like GetSyncer()

File: web/src/SyncerListPage.js [hunk 3]
```diff
@@ -205,12 +216,13 @@ class SyncerListPage extends BaseListPage {
...
         dataIndex: '',
         key: 'op',
-        width: '170px',
+        width: '240px',
         fixed: (Setting.isMobile()) ? "false" : "right",
         render: (text, record, index) => {
           return (
             <div>
-              <Button style={{marginTop: '10px', marginBottom: '10px', marginRight: '10px'}} type="primary" onClick={() => this.props.history.push(`/syncers/${record.name}`)}>{i18next.t("general:Edit")}</Button>
+              <Button style={{marginTop: '10px', marginBottom: '10px', marginRight: '10px'}} type="primary" onClick={() => this.syncUsers(index)}>{i18next.t("general:Sync")}</Button>
```
Comment (hsluoyz) [inline web/src/SyncerListPage.js:224]: Plz also add this button to the edit page

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/SyncerListPage.js (hunk 3 above)
Comment (seriouszyx) [reply web/src/SyncerListPage.js]: we have to write `runSyncer` function twice to do that, maybe it's now cleaner

File: controllers/syncer.go [hunk 4]
```diff
@@ -114,3 +115,19 @@ func (c *ApiController) DeleteSyncer() {
...
+// @Description run syncer
+// @Param   body    body   object.Syncer  true        "The details of the syncer"
+// @Success 200 {object} controllers.Response The Response object
+// @router /run-syncer [get]
+func (c *ApiController) RunSyncer() {
+	id := c.Input().Get("id")
+	syncer := object.GetSyncer(id)
+
+	object.RunSyncer(syncer)
+	time.Sleep(5 * time.Second)
```
Comment (hsluoyz) [inline controllers/syncer.go:130]: This is only for your test

File: controllers/syncer.go [hunk 5]
```diff
@@ -114,3 +115,19 @@ func (c *ApiController) DeleteSyncer() {
...
+
+	c.ResponseOk()
+}
```
Comment (hsluoyz) [inline controllers/syncer.go:133]: new line

File: object/syncer.go [hunk 6]
```diff
@@ -206,3 +206,8 @@ func (syncer *Syncer) getTable() string {
 		return syncer.Table
 	}
 }
+
+func RunSyncer(syncer *Syncer) {
+	syncer.initAdapter()
+	syncer.syncUsers()
+}
```
Comment (hsluoyz) [inline object/syncer.go:213]: new line

File: web/src/backend/SyncerBackend.js [hunk 7]
```diff
@@ -54,3 +54,10 @@ export function deleteSyncer(syncer) {
     body: JSON.stringify(newSyncer),
   }).then(res => res.json());
 }
+
+export function runSyncer(owner, name) {
+  return fetch(`${Setting.ServerUrl}/api/run-syncer?id=${owner}/${encodeURIComponent(name)}`, {
+    method: 'GET',
+    credentials: 'include',
+  }).then(res => res.json());
+}
```
Comment (hsluoyz) [inline web/src/backend/SyncerBackend.js:63]: new line

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/backend/SyncerBackend.js (hunk 7 above)
Comment (seriouszyx) [reply web/src/backend/SyncerBackend.js]: I added the new line but it seems git delete it automatically

image

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**